src/types/wrappers/xConstrainedAllocationModel.jl

```diff
@@ -0,0 +1,124 @@
+
+    ConstrainedAllocationModel(weight, growth_rate_id, total_enzyme_bound_id, inner) = begin
+        inner isa GeckoModel ||
+        inner isa SMomentModel ||
+        throw(error("Model type unsupported by ConstrainedAllocationModel."))
+        idx = first(indexin([growth_rate_id], reactions(inner)))
+        lbs, ubs = bounds(inner)
+        lbs[idx] * ubs[idx] < 0 &&
+        throw(error("Growth rate reaction must be unidirectional."))
+        new(weight, growth_rate_id, total_enzyme_bound_id, inner)
+    end
+end
+
+ConstrainedAllocationModel(weight::Float64, growth_rate_id::String, inner::SMomentModel) =
+    ConstrainedAllocationModel(weight, growth_rate_id, "", inner)
+
+"""
+$(TYPEDSIGNATURES)
+
+Get the internal [`GeckoModel`](@ref) or [`SMomentModel`](@ref) out of
+[`ConstrainedAllocationModel`](@ref).
+"""
+Accessors.unwrap_model(m::ConstrainedAllocationModel) = m.inner
+
+"""
+$(TYPEDSIGNATURES)
+
+Return the coupling of [`ConstrainedAllocationModel`](@ref). In addition to the
+coupling constraints of the underlying model, this adds the ribosome trade-off
+to the total proteome capacity bound.
+"""
+function Accessors.coupling(model::ConstrainedAllocationModel)
+    model.inner isa GeckoModel && return _cam_gecko(model)
+    model.inner isa SMomentModel && return _cam_smoment(model)
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+Helper function that adds the ribosome trade-off to a [`GeckoModel`](@ref).
+"""
+function _cam_gecko(model::ConstrainedAllocationModel)
+    gcols = gecko_column_reactions(model.inner)
+    innerC = coupling(model.inner.inner) * gcols
+    rxnC = gecko_reaction_coupling(model.inner)
+    enzcap = gecko_mass_group_coupling(model.inner)
+    fluxcaps =
+        spzeros(length(model.inner.coupling_row_mass_group), length(model.inner.columns))
+    row_idx = findfirst(
+        x -> x == model.total_enzyme_bound_id,
+        mg.group_id for mg in model.inner.coupling_row_mass_group
+    )
+
+    _idx = first(indexin([model.growth_rate_id], reactions(model.inner)))
+    isnothing(_idx) && throw(error("Growth rate reaction ID not found."))
+    # need to adjust idx using gmoment_column_reactions due to split reactions (assume biomass reaction is not reversible)
+    col_idx = first(findall(!iszero, gcols[_idx, :]))
+    fluxcaps[row_idx, col_idx] = model.weight
+
+    [
+        innerC spzeros(size(innerC, 1), n_genes(model))
+        rxnC spzeros(size(rxnC, 1), n_genes(model))
+        fluxcaps enzcap
+    ]
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+Helper function that adds the ribosome trade-off to a [`SMomentModel`](@ref).
+"""
+function _cam_smoment(model::ConstrainedAllocationModel)
+    _idx = first(indexin([model.growth_rate_id], reactions(model.inner)))
+    isnothing(_idx) && throw(error("Growth rate reaction ID not found."))
+    # need to adjust idx using smoment_column_reactions due to split reactions (assume biomass reaction is not reversible)
+    cols = smoment_column_reactions(model.inner)
```

Comment:
once we've got variables, can't this just be gathered from stoichiometry?
...also I should rename to _variables :]

Reply:
I don't think so, you need to know which column to adjust in the last row of the coupling matrix (that is where the total capacity constraint is). It's position depends on the expanded variables (which include the split forward and reverse reactions)